Add console.table support

nchevobbe opened the pull request:
This is taking me more time that I thought, I' putting this here to motivate me finish this.
This is by no mean review-ready, I really don't like how I populate table data.

I don't know if I should reuse the existing css class or create a new one
Also, I don't know if the fixture for ObjectClient is something we want

Fixes #65

linclark commented:
I haven't looked at it, but 👏🏻 👏🏻 👏🏻

linclark commented:
> I don't know if I should reuse the existing css class or create a new one

If it's easier to create new ones, then let's do that. Maybe prepending with `new-` , like `.new-consoletable`. Then we can search through and find anything that has `new-` in front of it when we get rid of the old console cruft.

> Also, I don't know if the fixture for ObjectClient is something we want

We probably will need that to mock it out for tests.

nt1m commented:
A few questions:
- does this support table header sorting like the old widget?
- any chance we can land a shared component on m-c so we can reuse it in the storage inspector?

nchevobbe commented:
> does this support table header sorting like the old widget?

No. For this PR, I'm aiming a basic implementation of console.table, which accurately displays the data. We might add things on top of that later.

> any chance we can land a shared component on m-c so we can reuse it in the storage inspector?

The current console does use a component which is used in the storage inspector too. I'm not aware of a migration of the storage inspector to React, but I think what we have in mind for the console.table is to have something quite simple. The current component allows many things, like sorting, column resizing, cell editing, ... I don't know if we want to have that level of functionnalities for console.table, and also. Furhtermore, I don't know, but I don't have much knowledge in React/Redux, how we could have a shared component and thus map its state in the redux app. I guess it's not trivial, but again, I might be wrong, and @linclark is better suited to answer such question.

linclark commented:
> but I think what we have in mind for the console.table is to have something quite simple

Yes, you are correct. There are multiple other React based tables already in m-c which would be more suitable for the storage component. This table is not meant as a generic table, at least not in the first iteration.

nchevobbe commented:
I just pushed an updated version of this PR (https://github.com/devtools-html/gecko-dev/pull/272/commits/edb02d1d2995b13561562551392270a562833a63). This is not finished yet (there are no tests, and I do not handle headers passed as a second argument), but I think the basics are here and the output seems correct to me.
@linclark and @bgrins , do you mind having a look at this and telling me what you think of this so I can iterate on your comments and add tests after that, thanks !

One thing I can already stress out : logging large table ( 1000 rows ) is much faster than in the old console 👍 (though I don't know if it's because of React or because there are way less markup and code in this version), but it really feels like an improvement to me 😀

bgrins commented:
> One thing I can already stress out : logging large table ( 1000 rows ) is much faster than in the old console 👍 (though I don't know if it's because of React or because there are way less markup and code in this version), but it really feels like an improvement to me 😀

Nice!   On first glance of the PR it looks good to me - I went through the examples on MDN and they seem to render about the same.

One thing I noticed missing when testing it out - there's a feature for restricting which columns are printed like so: https://developer.mozilla.org/en-US/docs/Web/API/Console/table#Restricting_the_columns_displayed

    function Person(firstName, lastName) {
      this.firstName = firstName;
      this.lastName = lastName;
    }

    var john = new Person("John", "Smith");
    var jane = new Person("Jane", "Doe");
    var emily = new Person("Emily", "Jones");

    console.table([john, jane, emily], ["firstName"]);

nchevobbe commented:
> One thing I noticed missing when testing it out - there's a feature for restricting which columns are printed

This should be good in https://github.com/devtools-html/gecko-dev/pull/272/commits/ea89666c77b4fc4295141836687278b4da95b412

linclark commented:
@nchevobbe Just want to let you know, I've started review on this. I have to step away for a bit, but expect to finish in the next hour or two.

linclark commented on devtools/client/webconsole/new-console-output/utils/messages.js:
```diff
@@ -81,6 +81,19 @@ function transformPacket(packet) {
+            !supportedClasses.includes(parameters[0].class)
+          ) {
+            // If the class of the first parameter is not supported,
+            // we handle the cole as a simple console.log
```
Nit: I don't think cole is the word you meant to put here.

nchevobbe replied:
I do not. 
DONE

linclark commented on devtools/client/webconsole/new-console-output/components/grip-message-body.js:
```diff
@@ -41,12 +41,14 @@ function GripMessageBody(props) {
       ? StringRep({
         object: grip,
-        useQuotes: false
+        useQuotes: false,
```
I'm realizing that we might want to move this out of here and pass it in. In the existing console, we use quotes around strings in tables.

nchevobbe commented on devtools/client/webconsole/new-console-output/components/table.js:
```diff
@@ -0,0 +1,189 @@
+
+    let property = data[index].value;
+
+    if (property.preview && property.preview.kind !== "ArrayLike") {
```
I don't really think so, or at least I can't picture how. I am not really doing any formatting stuff here, only retrieving the correct properties from the grip.

What I do acknowledge is that this function is not easy to read, I went through it a couple of times, and I'll try to iterate to have something simpler.

nchevobbe commented on devtools/client/webconsole/new-console-output/components/table.js:
```diff
@@ -0,0 +1,189 @@
+
+exports.Table = Table;
+exports.getParametersDataType = getParametersDataType;
+exports.getTableItems = getTableItems;
```
Yes, I wanted to exported them in order test them ( at least I think `getParametersDataType` deserve a whole suite of tests ).
I added the comments.

nchevobbe commented on devtools/client/webconsole/new-console-output/test/requireHelper.js:
```diff
@@ -30,5 +30,7 @@ requireHacker.global_hook("default", path => {
     case "Services":
     case "Services.default":
       return `module.exports = require("devtools/client/webconsole/new-console-output/test/fixtures/Services")`;
+    case "devtools/shared/client/main":
```
oh, bad copy/pasting here. It was meant to return the `ObjectClient` fixture , since `ObjectClient` is required from `devtools/shared/client/maindevtools/shared/client/main`

I'll make a separate commit for the tests when they are ready.

nchevobbe commented:
Thanks for the review ! I will fix what I can right away.

> For example, this doesn't handle the collections of compound types, but I think we can postpone that.

This shouldn't be difficult to implement. In fact, I think this isn't supported due to a specific test I added in the `getParametersDataType` function.

> I think the only thing we really need for this is tests. 

I agree, , I'll add some and update the PR.

linclark commented on devtools/client/webconsole/new-console-output/components/table.js:
```diff
@@ -0,0 +1,189 @@
+      return;
+    }
+
+    const client = new ObjectClient(jsterm.hud.proxy.client, parameters[0]);
```
We would probably move towards something more like the debugger, where we [pass in a client](https://github.com/devtools-html/debugger.html/blob/a6e7d2dbce15bac3efa24cddebd62f2222d3d709/public/js/actions/pause.js#L177) using props. This will make it possible for us to support Chrome RDP and Firefox RDP at the same time.

That's not important yet, but Servo will use Chrome RDP so would need to happen for that.

nchevobbe commented:
Okay so I tried to add some tests, and I'm having some trouble here.

1. I want to test that if I add a console.table message, it will dispatch the `messageTableDataGet` and  `messageTableDataReceive` actions.
The first one is not testable like the other action because it is dispatched in the `ConsoleTable componentWillMount` function.
The second one is dispatched from the first one. I found to many different things online to have a clear view of this, maybe you know something for this ?

2. In addition to the console.table packet et message stubs, I think we need to have the tableData stub, which is the console.table's parameter grip with all its properties. I started getting this in `browser_webconsole_update_stubs_console_api.js` , but it feels clunky as hell, as I duplicate many of the code that are in `componentWillMount` and `messageTableDataGet` . I don't know if we could use `messageTableDataGet` for example to get the grip from the mochitest.

You suggested to use a mochitest for now if I had trouble with the stubs, but I think we would have an issue too since the render of the table is async ( after we get the grip properties ), and we won't have a way to know when the data is ready.

I'm done for today, I jut wanted to write those down, maybe I'll have better ideas tomorrow.

linclark commented:
> You suggested to use a mochitest for now if I had trouble with the stubs, but I think we would have an issue too since the render of the table is async ( after we get the grip properties ), and we won't have a way to know when the data is ready.

I should have mentioned... there's a thing I'm doing in the new mochitests that could help with this. There is a generic `waitFor` function. You pass in a function, and it will test whether it returns a truthy value every 100 ms.

What you could do is pass in a function to `waitFor` that uses the DOM to figure out whether the data is ready. You could do this either by adding a class to the table that's only used for tests (something like `.table-data-ready`), or just checking whether there are DOM nodes with the expected text values.

Then you can have the function that you pass to `waitFor` return the DOM node of the table if you want to do more checks on the structure.

nchevobbe commented:
> Then you can have the function that you pass to waitFor return the DOM node of the table if you want to do more checks on the structure.

That will definitely help me writing a first mochitest version of the tests, thanks !

nchevobbe commented:
I'm currently writing the mochitest for this. In the same time, I pushed an update to the pr ( https://github.com/devtools-html/gecko-dev/pull/272/commits/9e1f0a6da6b77597a60e50850967f46b4b04a2c1 ) with the modifications you requested.
@linclark, @bgrins  could you have a look and tell me if it look fine ? More specifically, I'm now deferring the creation of the ConsoleOutput body after the proxy is ready ( https://github.com/devtools-html/gecko-dev/pull/272/files#diff-d264dddd34080899332b7f3bccb4f709R21 ), and I would like to know if it seems right. 
Thanks !

bgrins commented on devtools/client/webconsole/new-console-output/new-console-output-wrapper.js:
```diff
@@ -18,33 +18,35 @@ const store = configureStore();
-    openLink: (url) => {
-      owner.openLink(url);
-    },
+  jsterm.hud._initConnection().then(() => {
```
I wonder if we should instead be waiting for `connectionInited` in web console init before creating the NewConsoleOutputWrapper to avoid adding asynchrony in this constructor (aka pull the `if (this.NEW_CONSOLE_OUTPUT_ENABLED) {` out of initUI and into the `allReady` callback.

In particular, is there a case where dispatchMessageAdd could be called before the component is created?  This could cause races in tests that are returning DOM nodes right after dispatching.

bgrins replied:
We could alternatively add an init function on the NewConsoleOutputWrapper object that runs all this code and gets called during `allReady` so it's more explicit that it's not ready until init is called.

nchevobbe commented:
@linclark @bgrins this is ready for review.
I split the PR in multiple commits so it should be easier to review.
Like you suggested previously, I added a mochitest to test the console.table output.
I also added simple mocha tests for the action and the store.

linclark commented on devtools/client/webconsole/new-console-output/test/mochitest/test-console.html:
```diff
@@ -13,6 +13,14 @@
         }
       }
+
+      function doConsoleTable(data, constrainedHeaders = null) {
```
Instead of adding this to the existing `test-console.html`, let's add a `test-console-table.html`.

linclark commented on devtools/client/webconsole/new-console-output/test/mochitest/browser_webconsole_console_table.js:
```diff
@@ -0,0 +1,157 @@
+ * http://creativecommons.org/publicdomain/zero/1.0/ */
+
+"use strict";
+
```
It might be worth putting a comment up here that links to the [docs](https://developer.mozilla.org/en-US/docs/Web/API/Console/table) on `console.table`. Then there could be a comment above each test case that explains which of the cases it is handling.

nchevobbe replied:
sure

linclark commented on devtools/client/webconsole/new-console-output/test/mochitest/browser_webconsole_console_table.js:
```diff
@@ -0,0 +1,157 @@
+    nodes.push(node);
+  }
+
+  is(nodes.length, testCases.length,
```
I'm not sure whether this test is actually checking what it says it is. Would it make sense to do a `querySelectorAll` call here and get all of the `.message .new-consoletable` nodes, and then compare it to the length? This would make sure that there are no additional nodes.

nchevobbe replied:
Do you mean something like : 
```javascript
let consoleTableNodes = hud.ui.experimentalOutputNode.querySelectorAll(
  ".message .new-consoletable");

is(consoleTableNodes.length, testCases.length, 
  "console has the expected number of consoleTable items");
```
?

linclark replied:
yeah, that's what I'm thinking.

nchevobbe commented:
> This is looking great! Just a couple of minor changes. None of them would block me from committing, so let me know if any should happen as follow ups.

I'm going to do these in the next couple hours, they are quite small.

nchevobbe commented:
@linclark all should be good in https://github.com/devtools-html/gecko-dev/pull/272/commits/c4b0c871d3fa7a2a1d3b3d4d423ba4fd4be73d40

nchevobbe commented:
Rebased with https://github.com/devtools-html/gecko-dev/pull/310

linclark commented:
This is in 🎉🎉🎉🎉🎉

I had to do a little rebasing, so it landed with e393585 and the 3 commits before that.
